include/libhal-armcortex/interrupt.hpp

```diff
@@ -16,262 +16,310 @@
-     *
-     */
-    static constexpr uint32_t enable_mask_code = 0x1F;
+/**
+ * @brief Default hard fault handler
+ *
+ * Used by developers in debug mode to see if they landed in the default hard
+ * fault handler.
+ *
+ */
+void hard_fault_handler();
 
-    /**
-     * @brief Determines if the irq is within the range of ARM
-     *
-     * @return true - irq is enabled by default
-     * @return false - irq must be enabled to work
-     */
-    [[nodiscard]] constexpr bool default_enabled() const
-    {
-      return m_id < core_interrupts;
-    }
+/**
+ * @brief Default memory management fault handler
+ *
+ * Used by developers in debug mode to see if they landed in the default memory
+ * management fault handler.
+ *
+ */
+void memory_management_fault_handler();
 
-    [[nodiscard]] constexpr std::uint32_t to_irq_number() const
-    {
-      return static_cast<std::uint32_t>(m_id - core_interrupts);
-    }
+/**
+ * @brief Default bus fault handler
+ *
+ * Used by developers in debug mode to see if they landed in the default bus
+ * fault handler.
+ *
+ */
+void bus_fault_handler();
 
-    /**
-     * @brief the enable bit for this interrupt resides within one of the 32-bit
-     * registers within the "iser" and "icer" arrays. This function will return
-     * the index of which 32-bit register contains the enable bit.
-     *
-     * @return constexpr std::uint32_t - array index
-     */
-    [[nodiscard]] constexpr std::uint32_t register_index() const
-    {
-      return to_irq_number() >> index_position;
-    }
+/**
+ * @brief Default usage fault handler
+ *
+ * Used by developers in debug mode to see if they landed in the default usage
+ * fault handler.
+ *
+ */
+void usage_fault_handler();
 
-    /**
-     * @brief return a mask with a 1 bit in the enable position for this
-     * exception_number.
-     *
-     * @return constexpr std::uint32_t - enable mask
-     */
-    [[nodiscard]] constexpr std::uint32_t enable_mask() const
-    {
-      return 1U << (to_irq_number() & enable_mask_code);
-    }
+/**
+ * @brief Sets the interrupt vector table back to a form where it can be
+ * initialized again().
+ *
+ * Recommended to never call this function in your application. If it is called,
+ * it should be well documented why this choice was necessary.
+ *
+ * Will clear all pending interrupts and sets the internal state to
+ * uninitialized. This call is dangerous and should only be done before any
+ * drivers depend on their interrupts to function.
+ *
+ */
+void revert_interrupt_vector_table();
 
-    /**
-     * @brief Provides the index within the IVT
-     *
-     * @return constexpr size_t - the index position
-     */
-    [[nodiscard]] constexpr size_t vector_index() const
-    {
-      return m_id;
-    }
+/**
+ * @brief Initialize the interrupt vector table
+ *
+ * Using this function directly is not recommended. Use the templated version of
+ * this in drivers and in application code. Only use this if you need to control
```

Comment:
"Drivers and app code" Like, use the libhal-lpc40xx version as it may have custom rules for its IVT?

Reply:
All of the arm cortex MCUs obey the same rules when it comes to the IVT. There shouldn't be any variability outside of having a unique set interrupt request numbers.

Reply:
Oh I think I misunderstood you last time. And now I think I understand what you meant now. The number of interrupt is platform specific. Some may have only 16 some may have 55. Just depends on the device. So each driver that uses exceptions NEEDs to call `hal::cortex_m::initialize<hal::platform::irq::max>();` or call something like `hal::lpc40::initialize_interrupts()` which internally calls the previous function. That will do the work of creating a statically allocated array with the exact amount of memory needed to support all interrupts. Since the template instantiation will be the same for all drivers, they all reference the same memory, and thus no leakage. And since the initialize function only initializes if the input array doesn't match the one currently in the IVT, the multiple calls do nothing after the first call.